akka-stream/src/main/scala/akka/stream/impl/Sinks.scala

```diff
@@ -277,7 +277,10 @@ private[akka] class QueueSink[T]() extends GraphStageWithMaterializedValue[SinkS
             promise.failure(new IllegalStateException("You have to wait for previous future to be resolved to send another request"))
           case None ⇒
             if (buffer.isEmpty) currentRequest = Some(promise)
-            else sendDownstream(promise)
+            else {
+              if (buffer.used == maxBuffer - 1) tryPull(in)
```

Comment:
I hope I correctly understood the guard here. The buffer's `used` property might actually be equal to `maxBuffer`, but in this case, the last enqueued buffer element is either a `Success(None)` or `Failure(ex)` and the `tryPull` is not needed. Enqueuing of these special elements happens in the `onUpstreamFinish` or `onUpstreamFailure` callbacks, so the guard from `onPush` (`buffer.used < maxBuffer  - 1`) entails this guard here. (I found this condition somewhat confusing, until I understood that stream end markers are kept in the buffer together with regular elements sent by upstream, and that they are not enqueued in the `onPull` method).  

What I still don't understand is why we allocate a `maxBuffer + 1` sized buffer.  Which code path would lead to buffer having 17 elements in it?